File: 이성일/week4/여행경로.java
import java.util.*;

class Solution {

    static boolean[] visited;

    static String[][] tickets;

    static String[] answer;

    static boolean found;

    public String[] solution(String[][] tickets) {
        Solution.tickets = tickets;
        answer = new String[tickets.length + 1];
        visited = new boolean[tickets.length];
        answer[0] = "ICN";
        Arrays.sort(tickets, (o1, o2) -> (o1[0].equals(o2[0]) ? o1[1].compareTo(o2[1]) : o1[0].compareTo(o2[0])));
        dfs("ICN", 1);
        return answer;
    }

    static void dfs(String currentAirport, int cnt) {
        if (found)
            return;
        if (cnt == tickets.length + 1) {
            found = true;
            return;
        }
        for (int i = 0; i < tickets.length; i++) {
            if (visited[i])
                continue;
            if (!tickets[i][0].equals(currentAirport))
                continue;
            visited[i] = true;
            String temp = tickets[i][1];
            answer[cnt++] = temp;
            dfs(temp, cnt);
            if (found)
                return;
            cnt--;
            visited[i] = false;
        }
    }
}

File: 이성일/week4/여행경로.py
from collections import defaultdict
def solution(tickets):

    airports = defaultdict(list)
    length = len(tickets)
    
    for ticket in tickets:
        airports[ticket[0]].append(ticket[1])
        
    for k in airports.keys():
        airports[k].sort()
        
    stack = ["ICN"]
    path = []
    
    while stack:
        point = stack[-1]
        
        if len(airports[point]) == 0:
            path.append(stack.pop())
        else:
            stack.append(airports[point].pop(0))
    answer = path[::-1]
    
    return answer
